drivers/gpio/gpio_gecko.c

```diff
@@ -345,13 +345,13 @@ static void gpio_gecko_common_isr(void *arg)
 		port_dev = data->ports[i];
 		port_data = port_dev->driver_data;
 		enabled_int = int_status & port_data->pin_callback_enables;
-		int_status &= ~enabled_int;
-
-		gpio_fire_callbacks(&port_data->callbacks, port_dev,
-				     enabled_int);
+		if (enabled_int != 0) {
+			int_status &= ~enabled_int;
+			GPIO->IFC = enabled_int;
```

Comment:
I would add this before the for loop instead of this change:

`GPIO->IFC = int_status;`

It only writes once to IFC and even if something goes wrong (due to some bug) with `port_data->pin_callback_enables` the interrupt flags are cleared.

Or is there something i am missing?

Reply:
There's a subtlety: IMO we should only be clearing the flag register bit if it corresponds to an enabled interrupt that triggers a callback.  If the interrupt is disabled clearing its status from the flags discards the fact that an interrupt has occurred, which could cause loss of notification in a race condition.

That's just how I would do it.  We could instead even do `GPIO->IFC = (u32_t)-1` after we read the register before testing anything, and it would probably pass.  That's just doing more than we need, discarding information that somebody who's using the device to its full capability would have wanted preserved.

In the unlikely event something goes wrong (an enabled interrupt isn't cleared) I would rather the system kept re-entering the ISR, because that would be visible and allow the bug to be identified and fixed.

Reply:
Ok, i understand this reasoning, too.

Reply:
It is possible to have an interrupt for a pin configured without a callback attached, e.g. when the pin is used to wake up from a low power mode. The current proposal would work too but it's better to have `GPIO->IFC = int_status;` as proposed by @chrta. In the future `pin_callback_enables` will be removed.

Possibly we could do `GPIO->IFC = int_status & GPIO->IEN;` but I'm not convinced it's necessary. User application shouldn't interact directly with the hardware module which is 'own' by the driver.

Reply:
But `gpio_get_pending_int()` is used for detecting wakeup: wakeup doesn't trigger an interrupt or produce a callback.

Also AFAICT multiple drivers use the policy I've described where `pin_callback_enables` is used to filter.  Nobody's said anything that I've heard about that going away.

I don't plan to change this, but I can withdraw the PR if you want to do it another way.  If so, you should check the other drivers to make sure they behave in the way you want, and perhaps document the required behavior somewhere.

Reply:
> But `gpio_get_pending_int()` is used for detecting wakeup: wakeup doesn't trigger an interrupt or produce a callback.

Yes, that's a good point. Then we should have `GPIO->IFC = int_status & GPIO->IEN`.

> Also AFAICT multiple drivers use the policy I've described where `pin_callback_enables` is used to filter. Nobody's said anything that I've heard about that going away.

`pin_callback_enables` exists only since we have `gpio_pin_enable_callback`, `gpio_pin_disable_callback` functions. These are deprecated and soon will be removed. 

In general the approach of masking interrupt clear bits with `pin_callback_enables` is incorrect. The code in `gpio_gecko.c` driver will work well since `gpio_gecko_disable_callback` function clears both a bit in `pin_callback_enables` as well as interrupt enable bit. If however it didn't clear the interrupt enable bit a user application that calls `gpio_pin_disable_callback` risks getting stuck in the interrupt routine if an interrupt on a pin is triggered -> The ISR will be invoked but the bit responsible for triggering the interrupt will never be cleared. I believe quite a few of the drivers that use this policy have the problem. Since we are going to deprecate `gpio_pin_disable_callback`, `gpio_pin_enable_callback` functions we probably don't need to fix it.

> I don't plan to change this, but I can withdraw the PR if you want to do it another way. If so, you should check the other drivers to make sure they behave in the way you want, and perhaps document the required behavior somewhere.

It's OK, we can fix it later.